### DISCUSSION TITLE
What's the deal with CourseMap? Is it a course? Is it a map? Who knows!

### DISCUSSION THREAD

**Discussion by vabold** [riidefi/mkw#242] [General]
### Note
The goal of this discussion is to decide on logistical problems early, to avoid complications later on. Names used are subject to change.

## Intro
`CourseMap` is responsible for parsing and storing information about the KMP files found in course archives. Its static pointer is located at `0x809bd6e8` PAL. In the Ghidra server, it is mostly named `KmpHolder`, with some exceptions. Its structure consists of only pointers to classes classes called `Accessors`, labeled in the Ghidra as `KartpointHolder`, `EnemypathHolder`, etc. The only exception is the header structure at offset 0x4, which consists of the raw file and some parsed header data such as the version.

## Accessor Structure
Each class responsible for a KMP section has a base `Accessor` class. The structure is as follows:

```c++
class Accessor {
protected:
    // The object the accessor class handles
    // Instead of using "void**", we will use a template
    void** obj;
    // Decides how many objects will be in the pointer array
    u16 count;
    // Pointer to the beginning of the KMP section
    void* raw;
};
```

From there, each created class' pointers are set in `CourseMap::init`.

## File Structure
Due to multiple Accessor functions existing between `CourseMap` functions, we know they must exist in the same TU. This complicates things, as we cannot have each Accessor class in their own file. Instead, we must decide how the classes will be laid out.

One idea is to structure them similarly to `RaceConfig` and its subsidiary classes. As an example, `CheckpointAccessor` would exist as `CourseMap::CheckpointAccessor`, and the `Checkpoint` object would exist as `CourseMap::CheckpointAccessor::Checkpoint`.

## Additional Topics
During the discussion, new topics and new viewpoints will inevitably come up. This discussion post will be updated to reflect that.

**Comment by vabold** [comment]
> riidefi: `Holder` does not have ownership of the objects, `Accessor` is a better descriptor

riidefi is right - `Checkpoint` holds a pointer to the object directly in the KMP file, as does the Holder/Accessor. If necessary, the base class for an object should be called `Proxy`. I will update the discussion accordingly.

**Comment by em-eight** [comment]
### CourseMap boundaries
rodata:
start >= 0x8088f8e8 (do not duplicate float conversion constant)
start <= 0x8088fa24 (many data here is referenced in CourseMap constructor)
end >= 0x8088fa90 (string referenced in constructor)
end <= 0x8088faa4 (start of DvdArchive data)
I suggest splitting the smallest amount and we can always extend later
data:
start == 0x808b2c20 (things before that are Channelnstaller related and here is where CourseMap data start)
end == 0x808b2c78 (after that it's DvdArchive data and before that are KMP object vtables)
text:
start == 0x80512694 (CourseMap static instance creation, previous function references float conversion constant that would be duplicated if in CourseMap data)
end == 0x80518cc0 (after that it is DvdArchive text and before that it EnemyPoint destructor)
.bss
start == 0x809bd6e8 (CourseMap static instance)
end == 0x809bd6f8 (GameScene related stuff start)

**Reply by vabold** [reply]
> start == 0x80512694

0x80510d7c and forward is part of CourseMap as well.

**Comment by Thodoris1999** [comment]
It seems that `raw` is only used to access the section info (section name, count and additional value), so I think it is a pointer to such struct. It also makes sense because `obj` (suggesting to just call it `data`) can access the rest of the section data anyway.
